### PR TITLE
[devtools] Fix can't expand prop value in some scenario

### PR DESCRIPTION
**Pull request by iChenLei**
## Summary

**TL;DR**
1. A bug incoming with this PR #19786
2. Devtools extension user can't expand prop value in some scenario
3. Versions affected 4.9.0 / 4.10.0 / 4.10.1

## Bug Record Screenshot
As you can see, the mobxStores value not show correctly!
<div style="display:flex;justify-content:center;width=650px;">
		<p>
	    </p>
</div>
<div style="display:flex;justify-content:center;width=650px;">
		<p>
	    </p>
</div>
<div style="display:flex;justify-content:center;width=650px;">
		<p>
	    </p>
</div>

>  Don't focus the **legacy context** block, cause the same bug can reappear with context or hooks(useContext)

## Code Analysis
The code from pr #19786 by omarsy, [#L71-L97](https://github.com/facebook/react/blob/master/packages/react-devtools-shared/src/utils.js#L71-L97)
```js
export function getAllEnumerableKeys(
  obj: Object,
): Array<string | number | Symbol> {
    /*** omit ***/
    current = Object.getPrototypeOf(current);
   /*** omit ***/
}
```
[Mobx](https://github.com/mobxjs/mobx)(with React) user offen define data store this way:
```js
class ReactDataStore {
    @observable appStatus;

    @action.bonud
    changeAppStatus (status) {
        this.appStatus = status;
    }
}
```
Will transpile to: 
```js
/**** omit polyfill code ****/
class ReactDataStore {
    changeAppStatus(init) {
        this.appInit = init;
    }
}
__decorate([
    observable,
    __metadata("design:type", Object)
], ReactDataStore.prototype, "appInit", void 0);

__decorate([
    action.bonud,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], ReactDataStore.prototype, "changeAppStatus", null);

```
So when the react devtools start [dehydrate](https://github.com/facebook/react/blob/master/packages/react-devtools-shared/src/hydration.js#L102-L332) the react app instance data, the **cleaned** will create duplicate path cause the **getAllEnumerableKeys** function.
<div style="display:flex;justify-content:center;width=650px;">
		<p>
	    </p>
</div>

I set a *debugger* for react devtools panel page to find the reason of the bug.
### So what will happen when we create duplicate **cleaned** path? (💀  if the prop value is **undefined** )
```js
export function hydrate(
  object: any,
  cleaned: Array<Array<string | number>>,
  unserializable: Array<Array<string | number>>,
): Object {
  cleaned.forEach((path: Array<string | number>) => {
    const length = path.length;
    const last = path[length - 1];
    const parent = getInObject(object, path.slice(0, length - 1));
    if (!parent || !parent.hasOwnProperty(last)) {
      return;
    }
    const value = parent[last];

   if (value.type === 'infinity') {
      parent[last] = Infinity;
    } else if (value.type === 'nan') {
      parent[last] = NaN;
    } else if (value.type === 'undefined') {
      parent[last] = undefined;
    } else {
/*** omit  ***/
```
When we meet duplicate path firstly , the **parent[last]** is **{ type: undefined }**([data process logic](https://github.com/facebook/react/blob/master/packages/react-devtools-shared/src/backend/utils.js#L15-L39)), then **parent[last]** will be reassign as **undefined**。
```js
    } else if (value.type === 'undefined') {
      parent[last] = undefined;
    } else {
```
Everything is ok until we meet duplicate path secondly, this time our **parent[last]** is **undefined** not **{ type: undefined }**, cause the path is same. so error occured !
```js
Error in event handler: TypeError: Cannot read property 'type' of undefined
```
<div style="display:flex;justify-content:center;width=650px;">
		<p>
	    </p>
</div>

## Code Review 

<del>The PR #19786 is for *React DevTools: Show symbols used as keys in state*, but I change some code logic, so I hope @omarsy can review my PR,</del> of course @bvaughn is also important reviewer.
💔  Change the logic will broken devtools CI [detail](https://app.circleci.com/pipelines/github/facebook/react/8611/workflows/ac137a52-cab8-48a3-96e1-ded0ca7b7bf4/jobs/249856), so just skip **parent[last]** when the value is not a valid object.
```js
// previous commit to avoid duplicate cleaned path
// packages/react-devtools-shared/src/utils.js  -- getAllEnumerableKeys
- return keys;
+ return [...new Set(keys)];
```

Anyway, Thanks for bvaughn's great work for React Devtools. 👍 

## Test Plan

Necessary unit test will add in the this PR later,  if my pr is valuable.

### CODE REVIEW

File: packages/react-devtools-shared/src/hydration.js [hunk 1]
```diff
@@ -383,7 +383,9 @@ export function hydrate(
 
     const value = parent[last];
 
-    if (value.type === 'infinity') {
+    if (!value) {
+      return;
+    } else if (value.type === 'infinity') {
       parent[last] = Infinity;
     } else if (value.type === 'nan') {
```
Comment (omarsy) [inline packages/react-devtools-shared/src/hydration.js:390]: Maybe we can have same issue for Infinty ?

Comment (iChenLei) [reply packages/react-devtools-shared/src/hydration.js:390]: yeah, I miss the scenario which value is Infinity(Infinity is Truthy). Any suggestion for me ?  I think i did not understand your PR #19786 (I mean all logic detail) is the biggest obstacle to fix this bug. 
```js
- cleaned.forEach((path: Array<string | number>) => {
+ [...new Set(cleaned)].forEach((path: Array<string | number>) => {
```
Is these code ok? When we loop cleaned array, we filter the duplicate item and don't need check the **parent[last]**?

Comment (omarsy) [reply packages/react-devtools-shared/src/hydration.js:390]: We don't have duplicate keys here actually. In your example I think we have two symbols with the same name and as I have to change the symbols into a `string` so that we can do the `postmessage` that's why we have this issue.

**Comment by bvaughn** [comment]
This is a great PR description! Thank you for doing this detailed investigation and sharing your findings like this.

> Necessary unit test will add in the this PR later, if my pr is valuable.

I'd love to see a new unit tests or two for this, yes. This will help us avoid regressing in the future.

File: packages/react-devtools-shared/src/hydration.js (hunk 1 above)
Comment (iChenLei) [reply packages/react-devtools-shared/src/hydration.js:390]: I didn't write any **Symbol** in my business code,  Just **class(decorators)** 、**mobx**  and **react**. I think store and store prototype has same params is the reason. (mobx internal reactivity machine logic)
Mobx will add **@observable** params to the prototype, so getAllEnumerableKeys will return [..., 'finished', 'finished'].
> Look the codesandbox console or chrome devtools console
[Reproduce Codesandbox link](https://codesandbox.io/s/simple-mobx-todolist-forked-ysi4p) 
<div style="display:flex;justify-content:center;width=600px;">
		<p>
	    </p>
</div>

Comment (omarsy) [reply packages/react-devtools-shared/src/hydration.js:390]: Thank you. I can now reproduce the problem with a basic object.
[Reproduce issue]( https://codesandbox.io/s/simple-mobx-todolist-forked-jogl8?file=/index.js)
Maybe @bvaughn we should not include inherited keys or find a way to avoid duplicates ?
image

Comment (bvaughn) [reply packages/react-devtools-shared/src/hydration.js:390]: `p.p` in the above example only has one value (42) even though there are "duplicates". Our code (`getAllEnumerableKeys`) just needs to handle the fact that `Object.keys` and `Object.getOwnPropertyDescriptors` may overlap.

Seems like the easiest way would be to just use a `Set` internally?
```js
export function getAllEnumerableKeys(
  obj: Object,
): Array<string | number | Symbol> {
  const keys = new Set();
  let current = obj;
  while (current != null) {
    const currentKeys = [
      ...Object.keys(current),
      ...Object.getOwnPropertySymbols(current),
    ];
    const descriptors = Object.getOwnPropertyDescriptors(current);
    currentKeys.forEach(key => {
      // $FlowFixMe: key can be a Symbol https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor
      if (descriptors[key].enumerable) {
        keys.add(key);
      }
    });
    current = Object.getPrototypeOf(current);
  }
  return Array.from(keys);
}
```

Or maybe I'm overlooking something. 😄

Comment (bvaughn) [reply packages/react-devtools-shared/src/hydration.js:390]: Or maybe better yet just change the return type?

```diff
diff --git a/packages/react-devtools-shared/src/utils.js b/packages/react-devtools-shared/src/utils.js
index 2c796c222..e9ba55f5a 100644
--- a/packages/react-devtools-shared/src/utils.js
+++ b/packages/react-devtools-shared/src/utils.js
@@ -67,8 +67,8 @@ export function alphaSortKeys(
 
 export function getAllEnumerableKeys(
   obj: Object,
-): Array<string | number | Symbol> {
-  const keys = [];
+): Set<string | number | Symbol> {
+  const keys = new Set();
   let current = obj;
   while (current != null) {
     const currentKeys = [
@@ -79,7 +79,7 @@ export function getAllEnumerableKeys(
     currentKeys.forEach(key => {
       // $FlowFixMe: key can be a Symbol https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor
       if (descriptors[key].enumerable) {
-        keys.push(key);
+        keys.add(key);
       }
     });
     current = Object.getPrototypeOf(current);
@@ -728,7 +728,7 @@ export function formatDataForPreview(
       return data.toString();
     case 'object':
       if (showFormattedValue) {
-        const keys = getAllEnumerableKeys(data).sort(alphaSortKeys);
+        const keys = Array.from(getAllEnumerableKeys(data)).sort(alphaSortKeys);
 
         let formatted = '';
         for (let i = 0; i < keys.length; i++) {
```

Since we don't need an Array in some cases.

Comment (iChenLei) [reply packages/react-devtools-shared/src/hydration.js:390]: My previous commit is 
```js
 return [...new Set(keys)];
```
But it broken unit test [circle ci report](https://app.circleci.com/pipelines/github/facebook/react/8611/workflows/ac137a52-cab8-48a3-96e1-ded0ca7b7bf4/jobs/249856)... :cry:

Comment (bvaughn) [reply packages/react-devtools-shared/src/hydration.js:390]: I ran tests locally with the diff I posted in [my comment above](https://github.com/facebook/react/pull/20534#discussion_r551425609) and they all pass.

Comment (iChenLei) [reply packages/react-devtools-shared/src/hydration.js:390]: Ok, I rebase my PR with your code 😆 , and add unit test later. ( Now it's 02:16 in shanghai, china. I need to sleep. Thank u, bvaughn

Comment (bvaughn) [reply packages/react-devtools-shared/src/hydration.js:390]: Please rest! 🙇 

Ping me once you've added a test and I'll review again.

Comment (iChenLei) [reply packages/react-devtools-shared/src/hydration.js]: @bvaughn Hi, I had finished the unit test, please review, thank u very much ! 😆

Comment (bvaughn) [reply packages/react-devtools-shared/src/hydration.js]: Thanks. It's back in my queue.

Comment (iChenLei) [reply packages/react-devtools-shared/src/hydration.js]: If no problem in this pull request, I hope you can release a new small version (eg: 4.10.2) for react devtools , that's very helpful for my team. Thank you !  bvaughn

**Review by bvaughn — APPROVED** [review]

**Comment by bvaughn** [comment]
Soon as CI passes I'll land this. Thanks again for the great investigation work and for your patience.

**Comment by vetraz** [comment]
Thank you for your fixes! We faced the same problems and are waiting for this changes to be published.
@bvaughn can you shed some light on when it will be available in stores?
